hcl/hcl.go

```diff
@@ -1002,6 +816,255 @@ func blockIsAllowed(name string) bool {
 	}
 }
 
+func loadCfgBlocks(dir string) (*hclparse.Parser, error) {
+	logger := log.With().
+		Str("action", "loadCfgBlocks()").
+		Str("dir", dir).
+		Logger()
+
+	logger.Trace().Msg("listing files")
+
+	dirEntries, err := os.ReadDir(dir)
+	if err != nil {
+		return nil, fmt.Errorf("reading files to load terramate cfg: %v", err)
+	}
+
+	logger.Trace().Msg("looking for Terramate files")
+
+	parser := hclparse.NewParser()
+
+	for _, dirEntry := range dirEntries {
+		logger := logger.With().
+			Str("entryName", dirEntry.Name()).
+			Logger()
+
+		if dirEntry.IsDir() {
+			logger.Trace().Msg("ignoring dir")
+			continue
+		}
+
+		filename := dirEntry.Name()
+		if strings.HasSuffix(filename, ".tm") || strings.HasSuffix(filename, ".tm.hcl") {
+			logger.Trace().Msg("found Terramate config, reading file")
+
+			path := filepath.Join(dir, filename)
+			data, err := os.ReadFile(path)
+			if err != nil {
+				return nil, fmt.Errorf("reading Terramate config %q: %v", path, err)
+			}
+
+			logger.Trace().Msg("Terramate config read, parsing it")
+
+			_, diags := parser.ParseHCL(data, filename)
+			if diags.HasErrors() {
+				return nil, errutil.Chain(ErrHCLSyntax, diags)
+			}
+
+			logger.Trace().Msg("Terramate config file parsed successfully")
+		}
+	}
+
+	return parser, nil
+}
+
+func newCfgFromParsedHCLs(dir string, parser *hclparse.Parser) (Config, error) {
+	logger := log.With().
+		Str("action", "newCfgFromParsedHCLs()").
+		Str("dir", dir).
+		Logger()
+
+	tmconfig := Config{
+		absdir: dir,
+	}
+
+	for fname, hclfile := range parser.Files() {
+		logger := logger.With().
+			Str("filename", fname).
+			Logger()
+
+		body, ok := hclfile.Body.(*hclsyntax.Body)
+		if !ok {
+			return Config{}, fmt.Errorf("can't get hcl body of file %q", fname)
```

Comment:
We can do panic now that Im thinking about it, I ended up doing this because the current code also doens't panics immediately..it just omits the cast error:

https://github.com/mineiros-io/terramate/blob/7905b9e3f98b8e4e0aed1919d34b1580dd3700cf/hcl/hcl.go#L269
https://github.com/mineiros-io/terramate/blob/7905b9e3f98b8e4e0aed1919d34b1580dd3700cf/hcl/hcl.go#L187

This will cause a panic, but Im not even sure when or which panic it will be (execution goes on with a nil body).
So my instinct was to use the ommited result of the casting. But I could have just removed it, so it would panic on the cast.

Like this:

```go
body := val.(etc)
```

So it panics at the cast, WDYT ?

Reply:
yes this is better